File: 알고리즘 연습/조합순열/Excer.py
arr = [1,2,3,4]
# 순열
from itertools import permutations
result = list(permutations(arr,2))
print(result)

# 순열 구현(재귀)
result = []
def perm(result, current, arr, k) :
    if k==0 :
        result.append(current.copy()) # 깊은복사로 새로운 배열로 저장해야함
        return

    for i in range(len(arr)) :
        if arr[i] in current :
            continue
        current.append(arr[i])
        perm(result, current, arr, k-1)
        current.pop()
    
perm(result, [], arr, 2)
print(result)

# 중복순열 구현(재귀)
result = []
def perm(result, current, arr, k) :
    if k==0 :
        result.append(current.copy()) # 깊은복사로 새로운 배열로 저장해야함
        return

    for i in range(len(arr)) :
        # if arr[i] in current :
        #     continue
        current.append(arr[i])
        perm(result, current, arr, k-1)
        current.pop()
    
perm(result, [], arr, 2)
print(result)


# 조합
from itertools import combinations
result = list(combinations(arr, 2))
print(result)

# 조합 구현(재귀) 2번째
def comb(result, current, arr, start, k) :
    if k == 0:
        result.append(current.copy())
        return
    
    for i in range(start, len(arr)-k+1) :
        current.append(arr[i])
        comb(result, current, arr, i+1, k-1)
        current.pop()

result = []
comb(result, [], arr, 0, 2)
print(result)

# 중복가능 조합
def comb(result, current, arr, start, k) :
    if k == 0:
        result.append(current.copy())
        return
    
    for i in range (start, len(arr)-k+1):
        current.append(arr[i])
        comb(result, current, arr, 0, k-1)
        current.pop()

result = []
comb(result, [], arr, 0, 2)
print(result)

File: 알고리즘 연습/조합순열/Excer.java
import java.util.ArrayList;
import java.util.List;

public class Excer {
    public static void main(String[] args) {
        int[] arr = {1,2,3,4};
        List<List<Integer>> result = new ArrayList<>();
        perm(result, new ArrayList<>(), arr, 2);
        System.out.println(result);
    }

    // 조합
    static void comb(List<List<Integer>> result, List<Integer> current, int[] arr, int start, int k) {
        if (k == 0) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int i=start; i<=arr.length-k; i++) {
            current.add(arr[i]);
            comb(result, current, arr, i+1, k-1);
            current.remove(current.size()-1);
        }
    }

    // 순열
    static void perm(List<List<Integer>> result, List<Integer> current, int[] arr, int k) {
        if (k == 0) {
            result.add(new ArrayList<>(current));
            return;
        }

        for (int i=0; i<=arr.length-k; i++) {
            if (current.contains(arr[i])) continue;
            current.add(arr[i]);
            perm(result, current, arr, k-1);
            current.remove(current.size()-1);
        }
    }
}
